// react-select-css/es/AsyncCreatable.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import PropTypes from 'prop-types';
import React from 'react';
import Async from './Async';
import Creatable from './Creatable';
import Select from './Select';

var AsyncCreatableSelect =
/*#__PURE__*/
function (_React$Component) {
  _inherits(AsyncCreatableSelect, _React$Component);

  function AsyncCreatableSelect() {
    _classCallCheck(this, AsyncCreatableSelect);

    return _possibleConstructorReturn(this, _getPrototypeOf(AsyncCreatableSelect).apply(this, arguments));
  }

  _createClass(AsyncCreatableSelect, [{
    key: "focus",
    value: function focus() {
      this.select.focus();
    }
  }, {
    key: "render",
    value: function render() {
      var _this = this;

      return React.createElement(Async, this.props, function (_ref) {
        var ref = _ref.ref,
            asyncProps = _objectWithoutProperties(_ref, ["ref"]);

        var asyncRef = ref;
        return React.createElement(Creatable, asyncProps, function (_ref2) {
          var ref = _ref2.ref,
              creatableProps = _objectWithoutProperties(_ref2, ["ref"]);

          var creatableRef = ref;
          return _this.props.children(_objectSpread({}, creatableProps, {
            ref: function ref(select) {
              creatableRef(select);
              asyncRef(select);
              _this.select = select;
            }
          }));
        });
      });
    }
  }]);

  return AsyncCreatableSelect;
}(React.Component);

var defaultChildren = function defaultChildren(props) {
  return React.createElement(Select, props);
};

AsyncCreatableSelect.propTypes = {
  children: PropTypes.func.isRequired // Child function responsible for creating the inner Select component; (props: Object): PropTypes.element

};
AsyncCreatableSelect.defaultProps = {
  children: defaultChildren
};
export default AsyncCreatableSelect;
