// react-select-css/lib/AsyncCreatable.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _Async = _interopRequireDefault(require("./Async"));

var _Creatable = _interopRequireDefault(require("./Creatable"));

var _Select = _interopRequireDefault(require("./Select"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var AsyncCreatableSelect =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2.default)(AsyncCreatableSelect, _React$Component);

  function AsyncCreatableSelect() {
    (0, _classCallCheck2.default)(this, AsyncCreatableSelect);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(AsyncCreatableSelect).apply(this, arguments));
  }

  (0, _createClass2.default)(AsyncCreatableSelect, [{
    key: "focus",
    value: function focus() {
      this.select.focus();
    }
  }, {
    key: "render",
    value: function render() {
      var _this = this;

      return _react.default.createElement(_Async.default, this.props, function (_ref) {
        var ref = _ref.ref,
            asyncProps = (0, _objectWithoutProperties2.default)(_ref, ["ref"]);
        var asyncRef = ref;
        return _react.default.createElement(_Creatable.default, asyncProps, function (_ref2) {
          var ref = _ref2.ref,
              creatableProps = (0, _objectWithoutProperties2.default)(_ref2, ["ref"]);
          var creatableRef = ref;
          return _this.props.children(_objectSpread({}, creatableProps, {
            ref: function ref(select) {
              creatableRef(select);
              asyncRef(select);
              _this.select = select;
            }
          }));
        });
      });
    }
  }]);
  return AsyncCreatableSelect;
}(_react.default.Component);

var defaultChildren = function defaultChildren(props) {
  return _react.default.createElement(_Select.default, props);
};

AsyncCreatableSelect.propTypes = {
  children: _propTypes.default.func.isRequired // Child function responsible for creating the inner Select component; (props: Object): PropTypes.element

};
AsyncCreatableSelect.defaultProps = {
  children: defaultChildren
};
var _default = AsyncCreatableSelect;
exports.default = _default;
