// react-select-css/lib/utils/defaultMenuRenderer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var menuRenderer = function menuRenderer(_ref) {
  var focusedOption = _ref.focusedOption,
      focusOption = _ref.focusOption,
      inputValue = _ref.inputValue,
      instancePrefix = _ref.instancePrefix,
      onFocus = _ref.onFocus,
      onOptionRef = _ref.onOptionRef,
      onSelect = _ref.onSelect,
      optionClassName = _ref.optionClassName,
      optionComponent = _ref.optionComponent,
      optionRenderer = _ref.optionRenderer,
      options = _ref.options,
      removeValue = _ref.removeValue,
      selectValue = _ref.selectValue,
      valueArray = _ref.valueArray,
      valueKey = _ref.valueKey;
  var Option = optionComponent;
  return options.map(function (option, i) {
    var isSelected = valueArray && valueArray.some(function (x) {
      return x[valueKey] === option[valueKey];
    });
    var isFocused = option === focusedOption;
    var optionClass = (0, _classnames.default)(optionClassName, {
      'Select-option': true,
      'is-selected': isSelected,
      'is-focused': isFocused,
      'is-disabled': option.disabled
    });
    return _react.default.createElement(Option, {
      className: optionClass,
      focusOption: focusOption,
      inputValue: inputValue,
      instancePrefix: instancePrefix,
      isDisabled: option.disabled,
      isFocused: isFocused,
      isSelected: isSelected,
      key: "option-".concat(i, "-").concat(option[valueKey]),
      onFocus: onFocus,
      onSelect: onSelect,
      option: option,
      optionIndex: i,
      ref: function ref(_ref2) {
        onOptionRef(_ref2, isFocused);
      },
      removeValue: removeValue,
      selectValue: selectValue
    }, optionRenderer(option, i, inputValue));
  });
};

menuRenderer.propTypes = {
  focusOption: _propTypes.default.func,
  focusedOption: _propTypes.default.object,
  inputValue: _propTypes.default.string,
  instancePrefix: _propTypes.default.string,
  onFocus: _propTypes.default.func,
  onOptionRef: _propTypes.default.func,
  onSelect: _propTypes.default.func,
  optionClassName: _propTypes.default.string,
  optionComponent: _propTypes.default.func,
  optionRenderer: _propTypes.default.func,
  options: _propTypes.default.array,
  removeValue: _propTypes.default.func,
  selectValue: _propTypes.default.func,
  valueArray: _propTypes.default.array,
  valueKey: _propTypes.default.string
};
var _default = menuRenderer;
exports.default = _default;
